experimental/packages/opentelemetry-shim-opencensus/src/shim.ts

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { diag, trace, Tracer } from '@opentelemetry/api';
+import * as RequireInTheMiddle from 'require-in-the-middle';
+import type { Tracer as OcTracer, CoreTracer } from '@opencensus/core';
+
+import { ShimTracer } from './ShimTracer';
+import { VERSION } from './version';
+
+type CoreTracerConstructor = new (
+  ...args: ConstructorParameters<typeof CoreTracer>
+) => OcTracer;
+
+let hook: RequireInTheMiddle.Hooked | null = null;
+
+interface OpenCensusShimConfig {
+  /**
+   * An optional OpenTelemetry tracer to send OpenCensus spans to. If not provided, one will be
+   * created for you.
+   */
+  tracer?: Tracer | undefined;
+}
+
+/**
+ * Patches OpenCensus to redirect all instrumentation to OpenTelemetry. Uses
+ * require-in-the-middle to override the implementation of OpenCensus's CoreTracer.
+ *
+ * Use {@link uninstallShim} to undo the effects of this function.
+ *
+ * @param config
+ */
+export function installShim({
+  tracer = trace.getTracer('@opentelemetry/shim-opencensus', VERSION),
+}: OpenCensusShimConfig = {}): void {
+  diag.info('Installing OpenCensus shim require-in-the-middle hook');
+  hook = RequireInTheMiddle<{
```

Comment:
as far as I remember one reason to have the `@opentelemetry/instrumentation` package is to avoid that several modules do patching on their own and depend on RITM (and later import in the middle to get ESM support).
There is also some optimization/tuning there to use a single RITM hook for performance reasons.

Not sure if it is possible/useful but maybe this could be changed to use `@opentelemetry/instrumentation` instead RITM directly?

Reply:
Thanks, I wasn't aware of this.

> Not sure if it is possible/useful but maybe this could be changed to use `@opentelemetry/instrumentation` instead RITM directly?

I don't think it would make sense to expose the shim as an "instrumentation" conceptually that users need to put in `registerInstrumentations()`.

But, I suppose I could depend on `@opentelemetry/instrumentation` as an implementation detail which is not exposed as part of this module and just do `new MyPlugin().enable()` in the `installShim()` function. Wdyt?

Reply:
Yes, this is exactly what I had in mind.

Reply:
I gave it a shot but there seems to be a bug or some assumptions in `@opentelemetry/instrumentation` that prevents you from modifying a module export.  I'll try to make a minimal repro.

Reply:
So I think I found the issue.

1. The JS code that typescript generates for `@opencensus/core` uses a getter on the module.exports object to make properties immutable (see [`__createBinding()`](https://unpkg.com/@opencensus/core@0.1.0/build/src/index.js))
2. To get around that, I `return {...moduleExports, CoreTracer: Foo}` in the hook since I can't use `shimmer.wrap()` on an immutable property.
3. My InstrumentationNodeModuleDefinition`s [unpatch](https://github.com/open-telemetry/opentelemetry-js/blob/7255da994fe79f115ceb9b1260fb171cf9313706/experimental/packages/opentelemetry-instrumentation/src/platform/node/instrumentationNodeModuleDefinition.ts#L30) is expected to modify the *original* moduleExports to restore it but I never modified the original object. However, the RITM singleton hook continues to return the modified object I returned in step 2.

Reply:
I guess the solution would be to patch at the source of `CoreTracer` - the exports of `@opencensus/core/build/src/trace/model/tracer.js`. This should be a normal export.

Patching this one has the benefit that all imports of `CoreTracer` are effected (not only the reexport in index.js) - in case there are others.

Reply:
Resolved in #3809